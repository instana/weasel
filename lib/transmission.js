// @flow

import {encodeURIComponent, OriginalXMLHttpRequest} from './browser';
import {hasOwnProperty} from './util';
import type {Beacon} from './types';
import {info} from './debug';
import vars from './vars';

const maxLengthForImgRequest = 2000;

export function sendBeacon(data: Beacon) {
  const str = stringify(data);
  if (str.length === 0) {
    return;
  }

  if (DEBUG) {
    info('Transmitting beacon', data);
  }

  if (OriginalXMLHttpRequest && str.length > maxLengthForImgRequest) {
    const xhr = new OriginalXMLHttpRequest();
    xhr.open('POST', String(vars.reportingUrl), true);
    xhr.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
    xhr.send(str);
  } else {
    // Older browsers do not support the XMLHttpRequest API. This sucks and may
    // result in a variety of issues, e.g. URL length restrictions. "Luckily", older
    // browsers also lack support for advanced features such as resource timing.
    // This should make this transmission via a GET request possible.
    const image = new Image();
    image.src = String(vars.reportingUrl) + '?' + str;
  }
}

function stringify(data: Beacon) {
  let str = '';

  for (let key in data) {
    if (hasOwnProperty(data, key)) {
      const value = data[key];
      if (value != null) {
        str += '&' + encodeURIComponent(key) + '=' + encodeURIComponent(String(data[key]));
      }
    }
  }

  return str.substring(1);
}
