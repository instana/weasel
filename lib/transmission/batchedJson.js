// @flow

import { win, doc, nav, OriginalXMLHttpRequest } from '../browser';
import { addEventListener } from '../util';
import type { Beacon } from '../types';
import vars from '../vars';

const pendingBeacons: Array<Beacon> = [];
let pendingBeaconTransmittingTimeout;

export const isSupported =
  !!OriginalXMLHttpRequest &&
  !!win.JSON &&
  isVisibilityApiSupported() &&
  typeof vars.beaconBatchingTime === 'number' &&
  vars.beaconBatchingTime > 0 &&
  typeof nav.sendBeacon === 'function';

// We attempt batching of messages to be more efficient on the client, network and
// server-side. While the connection is either a persistent HTTP 2 connection or
// a HTTP 1.1 connection with keep-alive, there is still some overhead involved
// in having many small messages.
//
// For this reason we attempt batching. When batching we must be careful to
// force a transmission when the document is unloaded. To handle the variety
// of unload cases, we are making use of the Page Visibility API. This abstracts
// all unload cases.
if (isSupported) {
  addEventListener(doc, 'visibilitychange', function() {
    if (isWindowHidden()) {
      transmit();
    }
  });
}

export function sendBeacon(beacon: Beacon) {
  pendingBeacons.push(beacon);

  if (isWindowHidden()) {
    // We cannot guarantee that we will ever get time to transmit data in a batched
    // format, as this might occur while the document is being unloaded.
    // Immediately force a transmission.
    transmit();
    return;
  }

  // Once we reach this stage we know that we want to send out data in batches.
  // Schedule a new batch transmission only when there currently isn't a batch
  // transmission already scheduled.
  if (pendingBeaconTransmittingTimeout == null) {
    pendingBeaconTransmittingTimeout = setTimeout(transmit, vars.beaconBatchingTime);
  }
}

function transmit() {
  if (pendingBeaconTransmittingTimeout != null) {
    clearTimeout(pendingBeaconTransmittingTimeout);
    pendingBeaconTransmittingTimeout = null;
  }

  if (pendingBeacons.length === 0) {
    return;
  }

  const serializedBeacons = win.JSON.stringify(pendingBeacons);
  // clear the array
  pendingBeacons.length = 0;

  // This will transmit a text/plain;charset=UTF-8 content type. This is not what we
  // want, but changing the content type via the Blob constructor currently
  // breaks for cross-origin requests.
  // https://bugs.chromium.org/p/chromium/issues/detail?id=490015
  const sendBeaconState = nav.sendBeacon(String(vars.reportingUrl), serializedBeacons);

  // There are limits to the amount of data transmittable via the sendBeacon API.
  // If it doesn't work via the sendBeacon, try it via plain old AJAX APIs
  // as a last resort.
  if (sendBeaconState === false) {
    const xhr = new OriginalXMLHttpRequest();
    xhr.open('POST', String(vars.reportingUrl), true);
    xhr.setRequestHeader('Content-type', 'application/json;charset=UTF-8');
    // Ensure that browsers do not try to automatically parse the response.
    xhr.responseType = 'text';
    xhr.timeout = vars.xhrTransmissionTimeout;
    xhr.send(serializedBeacons);
  }
}

function isWindowHidden() {
  return doc.visibilityState === 'hidden';
}

function isVisibilityApiSupported() {
  return typeof doc.visibilityState === 'string';
}
