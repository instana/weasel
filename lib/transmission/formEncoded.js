// @flow

import {XMLHttpRequest, encodeURIComponent, executeImageRequest} from '../browser';
import { disableMonitoringForXMLHttpRequest } from '../hooks/xhrHelpers';
import {hasOwnProperty} from '../util';
import type {Beacon, ReportingBackend} from '../types';
import vars from '../vars';

const maxLengthForImgRequest = 2000;

export function sendBeacon(data: Beacon) {
  if (vars.reportingUrl && vars.reportingUrl.length > 0) {
    const str = stringify(data);
    if (str.length != 0) {
      transmit(str, String(vars.reportingUrl));
    }
    return;
  }

  if (!vars.reportingBackends || vars.reportingBackends.length === 0) {
    return;
  }

  for (let i = 0, len = vars.reportingBackends.length; i < len; i++) {
    let reportingBackend: ReportingBackend = vars.reportingBackends[i];
    if (hasOwnProperty(reportingBackend, 'reportingUrl')) {
      if (i > 0) {
        data['k'] = reportingBackend['key'];
      }
      const str = stringify(data);
      if (str.length != 0) {
        transmit(str, reportingBackend['reportingUrl']);
      }
    }
  }
}

function transmit(str: string, reportingUrl: string) {
  if (XMLHttpRequest && str.length > maxLengthForImgRequest) {
    const xhr = new XMLHttpRequest();
    disableMonitoringForXMLHttpRequest(xhr);
    xhr.open('POST', String(reportingUrl), true);
    xhr.setRequestHeader('Content-type', 'application/x-www-form-urlencoded;charset=UTF-8');
    // Ensure that browsers do not try to automatically parse the response.
    xhr.responseType = 'text';
    xhr.timeout = vars.xhrTransmissionTimeout;
    xhr.send(str);
  } else {
    // Older browsers do not support the XMLHttpRequest API. This sucks and may
    // result in a variety of issues, e.g. URL length restrictions. "Luckily", older
    // browsers also lack support for advanced features such as resource timing.
    // This should make this transmission via a GET request possible.
    executeImageRequest(String(reportingUrl) + '?' + str);
  }
}

function stringify(data: Beacon) {
  let str = '';

  for (let key in data) {
    if (hasOwnProperty(data, key)) {
      const value = data[key];
      if (value != null) {
        str += '&' + encodeURIComponent(key) + '=' + encodeURIComponent(String(data[key]));
      }
    }
  }

  return str.substring(1);
}
