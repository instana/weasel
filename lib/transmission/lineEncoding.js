// @flow

import {hasOwnProperty} from '../util';

// We know that must values are plain key/value pairs. We therefore choose a format that is
// very easy to parse in a streaming fashion on the server-side. This format is a basic
// line-based encoding of key/value pairs. Each line contains a key/value pair.
//
// In contrast to form encoding, this encoding handles JSON much better.
export function encode(beacons: Array<Object>) {
  let str = '';

  for (let i = 0; i < beacons.length; i++) {
    const beacon = beacons[i];

    // Multiple beacons are separated by an empty line
    str += '\n';

    for (let key in beacon) {
      if (hasOwnProperty(beacon, key)) {
        const value = beacon[key];
        if (value != null) {
          str += '\n' + encodePart(key) + '\t' + encodePart(value);
        }
      }
    }
  }

  return str.substring(2);
}

function encodePart(part: any) {
  return String(part)
    .replace(/\\/g, '\\\\')
    .replace(/\n/g, '\\n')
    .replace(/\t/g, '\\t');
}
