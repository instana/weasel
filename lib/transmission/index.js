// @flow

import { sendBeacon as sendBatchedJsonBeacon, isSupported as isBatchedJsonSupported } from './batchedJson';
import { sendBeacon as sendJsonBeacon, isSupported as isJsonSupported } from './json';
import {createExcessiveUsageIdentifier} from '../excessiveUsageIdentification';
import { sendBeacon as sendFormEncodedBeacon } from './formEncoded';
import type {Beacon} from '../types';
import {info, error} from '../debug';

export const jsonPayloadNativelySupported = isJsonSupported;

const isExcessiveUsage = createExcessiveUsageIdentifier({
  maxCalls: 1024,
  maxCallsPerTenSeconds: 128
});

export function sendBeacon(data: Beacon) {
  if (DEBUG) {
    info('Transmitting beacon', data);
  }

  if (isExcessiveUsage()) {
    if (DEBUG) {
      info('Reached the maximum number of beacons to transmit.');
    }
    return;
  }

  try {
    if (isBatchedJsonSupported) {
      sendBatchedJsonBeacon(data);
    } else if (isJsonSupported) {
      sendJsonBeacon(data);
    } else {
      sendFormEncodedBeacon(data);
    }
  } catch (e) {
    if (DEBUG) {
      error('Failed to transmit beacon', e);
    }
  }
}
