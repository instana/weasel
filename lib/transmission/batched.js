// @flow

import { doc, nav, OriginalXMLHttpRequest, sendBeacon as sendBeaconInternal } from '../browser';
import { addEventListener } from '../util';
import { encode } from './lineEncoding';
import type { Beacon } from '../types';
import vars from '../vars';

const pendingBeacons: Array<Beacon> = [];
let pendingBeaconTransmittingTimeout;

const isVisibilityApiSupported = typeof doc.visibilityState === 'string';
const isSupported = !!OriginalXMLHttpRequest && isVisibilityApiSupported && isSendBeaconApiSupported();

export function isEnabled() {
  return isSupported && vars.beaconBatchingTime > 0;
}

// We attempt batching of messages to be more efficient on the client, network and
// server-side. While the connection is either a persistent HTTP 2 connection or
// a HTTP 1.1 connection with keep-alive, there is still some overhead involved
// in having many small messages.
//
// For this reason we attempt batching. When batching we must be careful to
// force a transmission when the document is unloaded. To handle the variety
// of unload cases, we are making use of the Page Visibility API. This abstracts
// all unload cases.
if (isSupported) {
  addEventListener(doc, 'visibilitychange', function() {
    if (isWindowHidden()) {
      transmit();
    }
  });
}

export function sendBeacon(beacon: Beacon) {
  pendingBeacons.push(beacon);

  // We cannot guarantee that we will ever get time to transmit data in a batched
  // format when the window is hidden, as this might occur while the document is
  // being unloaded. Immediately force a transmission in these cases.
  if (!isWindowHidden() && vars.beaconBatchingTime > 0) {
    if (pendingBeaconTransmittingTimeout == null) {
      pendingBeaconTransmittingTimeout = setTimeout(transmit, vars.beaconBatchingTime);
    }
  } else {
    transmit();
  }
}

function transmit() {
  if (pendingBeaconTransmittingTimeout != null) {
    clearTimeout(pendingBeaconTransmittingTimeout);
    pendingBeaconTransmittingTimeout = null;
  }

  if (pendingBeacons.length === 0) {
    return;
  }

  const serializedBeacons = encode(pendingBeacons);
  // clear the array
  pendingBeacons.length = 0;

  // Empty beacons. Should never happen, but better be safe.
  if (serializedBeacons.length === 0) {
    return;
  }

  // This will transmit a text/plain;charset=UTF-8 content type. This may not be what we
  // want, but changing the content type via the Blob constructor currently
  // breaks for cross-origin requests.
  // https://bugs.chromium.org/p/chromium/issues/detail?id=490015
  const sendBeaconState = isSendBeaconApiSupported() && sendBeaconInternal(String(vars.reportingUrl), serializedBeacons);

  // There are limits to the amount of data transmittable via the sendBeacon API.
  // If it doesn't work via the sendBeacon, try it via plain old AJAX APIs
  // as a last resort.
  if (sendBeaconState === false) {
    const xhr = new OriginalXMLHttpRequest();
    xhr.open('POST', String(vars.reportingUrl), true);
    xhr.setRequestHeader('Content-type', 'text/plain;charset=UTF-8');
    // Ensure that browsers do not try to automatically parse the response.
    xhr.responseType = 'text';
    xhr.timeout = vars.xhrTransmissionTimeout;
    xhr.send(serializedBeacons);
  }
}

function isWindowHidden() {
  return doc.visibilityState === 'hidden';
}

function isSendBeaconApiSupported() {
  return typeof nav.sendBeacon === 'function';
}
