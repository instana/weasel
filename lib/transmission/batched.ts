import { doc, nav, XMLHttpRequest, sendBeacon as sendBeaconInternal } from '../browser';
import { disableMonitoringForXMLHttpRequest } from '../hooks/xhrHelpers';
import {onLastChance} from '../events/onLastChance';
import {setTimeout, clearTimeout} from '../timers';
import { encode } from './lineEncoding';
import type { Beacon, ReportingBackend } from '../types';
import { warn } from '../debug';
import vars from '../vars';

const maxBatchedBeacons = 15;

const pendingBeacons: Array<Partial<Beacon>> = [];
let pendingBeaconTransmittingTimeout: ReturnType<typeof setTimeout> | null;

const isVisibilityApiSupported = typeof doc.visibilityState === 'string';
const isSupported = !!XMLHttpRequest && isVisibilityApiSupported && isSendBeaconApiSupported();

export function isEnabled() {
  return isSupported && vars.beaconBatchingTime > 0;
}

// We attempt batching of messages to be more efficient on the client, network and
// server-side. While the connection is either a persistent HTTP 2 connection or
// a HTTP 1.1 connection with keep-alive, there is still some overhead involved
// in having many small messages.
//
// For this reason we attempt batching. When batching we must be careful to
// force a transmission when the document is unloaded.
if (isSupported) {
  onLastChance(transmit);
}

export function sendBeacon(beacon: Partial<Beacon>) {
  pendingBeacons.push(beacon);

  if (pendingBeacons.length >= maxBatchedBeacons) {
    transmit();
  } else if (!isWindowHidden() && vars.beaconBatchingTime > 0) {
    // We cannot guarantee that we will ever get time to transmit data in a batched
    // format when the window is hidden, as this might occur while the document is
    // being unloaded. Immediately force a transmission in these cases.
    if (pendingBeaconTransmittingTimeout == null) {
      pendingBeaconTransmittingTimeout = setTimeout(transmit, vars.beaconBatchingTime);
    }
  } else {
    transmit();
  }
}

function transmit() {
  if (pendingBeaconTransmittingTimeout != null) {
    clearTimeout(pendingBeaconTransmittingTimeout);
    pendingBeaconTransmittingTimeout = null;
  }

  if (pendingBeacons.length === 0) {
    return;
  }

  if (vars.reportingBackends && vars.reportingBackends.length > 0) {
    for (let i = 0, len = vars.reportingBackends.length; i < len; i++) {
      const reportingBackend: ReportingBackend = vars.reportingBackends[i];
      if (i > 0) {
        for (let j = 0, length = pendingBeacons.length; j < length; j++) {
          const beacon: Partial<Beacon> = pendingBeacons[j];
          beacon['k'] = reportingBackend['key'];
        }
      }
      transmitBeacons(encode(pendingBeacons), reportingBackend['reportingUrl']);
    }
  } else {
    transmitBeacons(encode(pendingBeacons), String(vars.reportingUrl));
  }

  // clear the array
  pendingBeacons.length = 0;
}

function transmitBeacons(serializedBeacons: string, reportingUrl: string) {
  // Empty beacons. Should never happen, but better be safe.
  if (!serializedBeacons || serializedBeacons.length === 0 || !reportingUrl || reportingUrl.length === 0) {
    return;
  }

  let sendBeaconState = false;
  if (isSendBeaconApiSupported()) {
    try {
      // This will transmit a text/plain;charset=UTF-8 content type. This may not be what we
      // want, but changing the content type via the Blob constructor currently
      // breaks for cross-origin requests.
      // https://bugs.chromium.org/p/chromium/issues/detail?id=490015
      sendBeaconState = sendBeaconInternal(String(reportingUrl), serializedBeacons);
    } catch (e) {
      // We have received reports that the navigator.sendBeacon API is failing in certain
      // Edge 14.x versions. Unfortunately, we cannot reproduce this with our existing
      // testing facilities. So we add a try/catch with logging and XMLHttpRequest
      // as a fallback.
      if (DEBUG) {
        warn(
          'navigator.sendBeacon has thrown an unexpected error. Will fall back to other transmission strategy.',
          'navigator.sendBeacon parameters:',
          String(vars.reportingUrl),
          serializedBeacons
        );
      }
    }
  }

  // There are limits to the amount of data transmittable via the sendBeacon API.
  // If it doesn't work via the sendBeacon, try it via plain old AJAX APIs
  // as a last resort.
  if (sendBeaconState === false) {
    const xhr = new XMLHttpRequest();
    disableMonitoringForXMLHttpRequest(xhr);
    xhr.open('POST', String(reportingUrl), true);
    xhr.setRequestHeader('Content-type', 'text/plain;charset=UTF-8');
    // Ensure that browsers do not try to automatically parse the response.
    xhr.responseType = 'text';
    xhr.timeout = vars.xhrTransmissionTimeout;
    xhr.send(serializedBeacons);
  }
}

function isWindowHidden() {
  return doc.visibilityState !== 'visible';
}

function isSendBeaconApiSupported() {
  return typeof nav.sendBeacon === 'function';
}
