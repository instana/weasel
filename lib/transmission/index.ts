// @flow

import { sendBeacon as sendBatchedBeacon, isEnabled as isBatchingEnabled } from './batched';
import {createExcessiveUsageIdentifier} from '../excessiveUsageIdentification';
import { sendBeacon as sendFormEncodedBeacon } from './formEncoded';
import {isUrlIgnored} from '../ignoreRules';
import type {Beacon} from '../types';
import {info, error} from '../debug';

const isExcessiveUsage = createExcessiveUsageIdentifier({
  maxCalls: 8096,
  maxCallsPerTenMinutes: 4096,
  maxCallsPerTenSeconds: 128
});

export function sendBeacon(data: Partial<Beacon>) {
  if (isUrlIgnored(data['l'])) {
    // data['l'] is a standardized property across all beacons to ensure that we do not accidentally transmit data
    // about a page such as this.

    if (DEBUG) {
      info('Skipping transmission of beacon because document URL associated to the beacon is ignored by ignore rule.', data);
    }
    return;
  }

  if (DEBUG) {
    info('Transmitting beacon', data);
  }

  if (isExcessiveUsage()) {
    if (DEBUG) {
      info('Reached the maximum number of beacons to transmit.');
    }
    return;
  }

  try {
    if (isBatchingEnabled()) {
      sendBatchedBeacon(data);
    } else {
      sendFormEncodedBeacon(data);
    }
  } catch (e) {
    if (DEBUG) {
      error('Failed to transmit beacon', e);
    }
  }
}
