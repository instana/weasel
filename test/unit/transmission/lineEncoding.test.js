import {encode} from '../../../lib/transmission/lineEncoding';

describe('transmission/lineEncoding', () => {
  it('must encode just one beacon', () => {
    const beacons = [
      {
        page: 'landing',
        key: 'abcdefg',
        id: 42,
        pageVisible: true
      }
    ];

    expect(encode(beacons)).toMatchSnapshot();
  });

  it('must encode multiple beacons', () => {
    const beacons = [
      {
        page: 'landing',
        key: 'abcdefg',
        id: 42,
        pageVisible: true
      },
      {
        page: 'product-details',
        key: 'abcdefg',
        id: 43,
        pageVisible: false,
        m_spec: JSON.stringify({signedIn: true})
      }
    ];

    expect(encode(beacons)).toMatchSnapshot();
  });

  it('must handle reserved characters in keys and values', () => {
    const beacons = [
      {
        'key\nwith\nnew\nline': 'a',
        'key\twith\ttabstop': 'b',
        'value with new line': 'a\nb\nc',
        'value with tabstop': 'a\tb\tc',
        'key\\with\nall\treserved\\ncharacters': 'value\\ with\nall\treserved\\ncharacter',
      }
    ];
    expect(encode(beacons)).toMatchSnapshot();
  });
});
