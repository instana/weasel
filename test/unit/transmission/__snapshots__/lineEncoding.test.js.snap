// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transmission/lineEncoding must encode just one beacon 1`] = `
"page	landing
key	abcdefg
id	42
pageVisible	true"
`;

exports[`transmission/lineEncoding must encode multiple beacons 1`] = `
"page	landing
key	abcdefg
id	42
pageVisible	true

page	product-details
key	abcdefg
id	43
pageVisible	false
m_spec	{\\"signedIn\\":true}"
`;

exports[`transmission/lineEncoding must handle reserved characters in keys and values 1`] = `
"key\\\\nwith\\\\nnew\\\\nline	a
key\\\\twith\\\\ttabstop	b
value with new line	a\\\\nb\\\\nc
value with tabstop	a\\\\tb\\\\tc
key\\\\\\\\with\\\\nall\\\\treserved\\\\\\\\ncharacters	value\\\\\\\\ with\\\\nall\\\\treserved\\\\\\\\ncharacter"
`;
